[Tan Wei Lie Eugene] iP

Eugenetanwl3881 opened the pull request:
# Bob: Your Personal Note-Taking Assistant :construction_worker:

> "Sometimes I'll start a sentence and I don't even know where it's going. I just hope I find it along the way." -  [Michael Scott](https://en.wikipedia.org/wiki/Michael_Scott_(The_Office))  

Bob frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

## How do I get it?
All you need to do is,
1. download it from [here](https://github.com/Eugenetanwl3881/ip/releases/tag/A-Jarl).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE!**

## What can I do?
Features:
- [x] Managing tasks
- [ ]  Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

## Practice Java
If you Java programmer, you can use it to practice Java too. Please ensure you have '/data/duke.txt' in your home directory
 Here's the main method:
```
    public static void main(String[] args) {
        String home = System.getProperty("user.home");
        new Duke(home + "/data/duke.txt").run();
    }
```

## Commands
Here are the commands available:
| Command  | Action |
| ------------- | ------------- |
| Bye  | Exits program  |
| Todo  | Creates a ToDo  |
| Deadline /by _date_  | Creates a deadline |
| Event /at _date_  | Creates an event |
| List  | List all tasks |
| Find _word_  | Finds task with given _word_ |
| Mark _task number_ | Marks task as done |
| Unmark _task number_  | Marks task as undone |

Commit:
Add Gradle support

Eugenetanwl3881 committed:
Complete Level 1

Add the input and output for the chatbot

Eugenetanwl3881 committed:
Complete Level 2

Used array to store all the user input

Eugenetanwl3881 committed:
Add Task class

Task class encapsulates a task and can determine/change whether task is done or not

Eugenetanwl3881 committed:
Complete Level 3

Let's Add the ability to mark and unmark task using an ArrayList of
tasks

Eugenetanwl3881 committed:
Complete Level 4

Let's make Task an abstract class so we can use inheritance for the
different other task types: ToDo, Deadline and Event. We create these
subtypes and their respective behaviors which will be respectively
called due to polymorphism

Eugenetanwl3881 committed:
Add Automated Text UI Testing

Update input.txt and EXPECTED.txt

Eugenetanwl3881 committed:
Complete Level 5

Let's add some exception handling deal with errors such as incorrect
inputs entered by the user. We will add two exceptions, incorrect input
and incomplete input

Eugenetanwl3881 committed:
Complete Level 6

Add the functionality to delete task in the list

Eugenetanwl3881 committed:
Implement A-Enums

Refactored code to use enum Command class

Eugenetanwl3881 committed:
Add A-Ecxceptions

Let's add more error-handling for user inputs

Eugenetanwl3881 committed:
Change print output for cases

Eugenetanwl3881 committed:
Rerun Automated Text UI Testing

Eugenetanwl3881 committed:
Fix Automated Text UI Test expected output

Eugenetanwl3881 committed:
Add Storage class

This is an initial implementation of the storage class

Let's
* use a storage class to read and write tasks into the storage

Eugenetanwl3881 committed:
Reorganize files into packages for future developement

Let's
* Reorganzie files into packages
* Add a TaskList class
* Add a UI class
* Add a RequestHandler class to deal with enum for requests

Eugenetanwl3881 committed:
Reorganize files into packages for future developement (Part 2)

Let's
* Reorganzie files into packages
* Add a TaskList class
* Add a UI class
* Add a RequestHandler class to deal with enum for requests

Eugenetanwl3881 committed:
Merge branch 'ReorganizeFiles' into branch-Level-7

Eugenetanwl3881 committed:
Add Save for Level 7

Saving and loading tasks from a file allows users to retrieve their
files

Let's
*save the task into a file whenever a change is made
*load previous saved file when running application
*reorganize packages for more OOP

Eugenetanwl3881 committed:
Merge branch 'branch-Level-7' into branch-Level-8

Eugenetanwl3881 committed:
Add Level 8: Dates and Time

Users can now add deadlines and event that are using date and time
format

Let's have the variables that are related to dates in event and
deadlines to be in LocalDate format

Eugenetanwl3881 committed:
Merge branch 'branch-Level-7'

Eugenetanwl3881 committed:
Merge branch 'branch-Level-8'

Eugenetanwl3881 committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

Eugenetanwl3881 committed:
Run application using Gradle

Update the build.gradle file to allow for the application to run using
Gradle

Eugenetanwl3881 committed:
Set up checkstyle for Gradle

Set up the necessary files for checkstyle as per Gradle tutorial@SE-EDU

Eugenetanwl3881 committed:
Add JUnit Tests

Eugenetanwl3881 committed:
Add Level 9 Find

Let's add the find command to find tasks by keywords

Eugenetanwl3881 committed:
Add JavaDoc

Let's add JavaDoc for all public classes and methods

Eugenetanwl3881 committed:
Add JavaDoc for classes/method used to implement Find

Eugenetanwl3881 committed:
Tweak code to follow Coding Standards

Eugenetanwl3881 committed:
Merge branch 'branch-Level-9'

Eugenetanwl3881 committed:
Merge branch 'branch-A-JavaDoc'

Merge 'branch-A-JavaDoc' and resolve merge conflicts in
src/main/java/duke/TaskList.java

Eugenetanwl3881 committed:
Merge branch 'branch-A-CodingStandard'

Let's tidy up the code to follow coding standards. Resolved any merge
conflicts as well.

Eugenetanwl3881 committed:
Update CheckStyle errors

Eugenetanwl3881 committed:
Add GUI using javafx

Eugenetanwl3881 committed:
Fix CheckStyle errors

Eugenetanwl3881 committed:
Merge branch 'branch-A-CheckStyle' into branch-A-Varargs

Eugenetanwl3881 committed:
Remove unused methods and update JavaDoc

Eugenetanwl3881 committed:
Use Varargs

Eugenetanwl3881 committed:
Merge branch 'branch-Level-10'

Eugenetanwl3881 committed:
Merge branch 'branch-A-CheckStyle'

Eugenetanwl3881 committed:
Merge branch 'branch-A-Varargs'

Kang-Quan commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,90 @@
+ * The Ui class handles the output to be printed by Bob that is shown to the user.
+ */
+public class Ui {
+    private static final String horizontalLine = "_____________________________________________________________";
```
I like how you have abstracted out `horizontalLine` as a static variable to reduce code duplication.

Kang-Quan commented on src/main/java/duke/DialogBox.java:
```diff
@@ -0,0 +1,64 @@
+import javafx.scene.text.Text;
+import javafx.scene.text.TextAlignment;
+
+/**
```
I like how most of your public classes and methods have JavaDocs comments making it easier to read your code. 👍

Kang-Quan commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,62 @@
+    @Override
+    public String saveStringFormat() {
+        return String.format("D | %d | %s | %s", this.isDone ? 1 : 0, this.description,
+                this.returnBy.format(DateTimeFormatter.ofPattern("MMM d yyyy")));
```
To reduce code duplication, perhaps formatting LocalDate to String can be further abstracted into a method so that in line 59, we do not have to repeat the code. This issue can also be seen in the Event class too.

Kang-Quan reviewed (commented):
Overall, I found your code easy to read for the most part. Most public classes have JavaDocs comments making it easy to understand. However, there were some duplication of code that can be reduced by abstracting them out into method if that make sense.

ama-chi commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,187 @@
+     * @return String representation of marked task.
+     */
+    public String markTask(int index) {
+        this.taskList.get(index - 1).markDone();
+        return this.taskList.get(index - 1).toString();
```
"this" may not be necessary here

ama-chi commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,62 @@
+     */
+    @Override
+    public String saveStringFormat() {
+        return String.format("D | %d | %s | %s", this.isDone ? 1 : 0, this.description,
```
For easier amendments in the future, the formats could perhaps been placed into a variable, so that if you need to adjust the format, you will only need to edit that variable.

ama-chi commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,187 @@
+        if (this.taskList.size() == 0) {
+            return "There are no tasks in your list";
+        } else {
+            StringBuilder str = new StringBuilder(String.format("There %s %d Bob task(s) in your list:",
```
I like the use of StringBuilder here to create your list

ama-chi reviewed (commented):
Overall, your structure is very clean and there is minimal repeating codes, good job!

Eugenetanwl3881 committed:
Add Assertions to Bob Chatbot

There are methods in the TaskList that requires valid input else they
will fail. There are also return values that required to be not null
else the program will fail.

Using Assertions, we can define assumptions about the program state at
given points so that the runtime can verify them.

Let's add assertions that should hold throughout various points of the
code, especially in TaskList and Duke.
Let's

Eugenetanwl3881 committed:
Improve Code Quality

Certain parts of current code do not follow good code quality. There
are also unused code that should be removed.

We can refactor the code to make it more understandable to readers.

Let's
* Refactor the code
* Remove unnecessary code

Eugenetanwl3881 committed:
Add Streams

We are iterating through an ArrayList to get the task and matching them
with the keyword.

We can use streams to simplify this.

Let's use streams to filter out the tasks based on the wordsToFind

Eugenetanwl3881 committed:
Merge pull request #1 from Eugenetanwl3881/branch-A-Assertions

Add Assertions to Bob Chatbot

Eugenetanwl3881 committed:
Merge pull request #2 from Eugenetanwl3881/branch-A-CodeQuality

Improve Code Quality

Eugenetanwl3881 committed:
Merge branch 'master' into branch-A-Streams

Eugenetanwl3881 committed:
Merge pull request #3 from Eugenetanwl3881/branch-A-Streams

Add Streams

Eugenetanwl3881 committed:
Add Update Extension

Current bot does not allow for updating of the task's description and
date without deleting and adding.

Giving user's the functionality to update will make the bot more useful.

Let's
* Add the command to update a task's description
* Add the command to update a task's date

Eugenetanwl3881 committed:
Merge branch 'branch-BCD-Extension'

Eugenetanwl3881 committed:
Add CI

Let's set up continuous integration by adding the gradle.yml file

Eugenetanwl3881 committed:
Merge branch 'branch-CI'

Eugenetanwl3881 committed:
Fix Checkstyle errors

Let's fix the checkstyle errors in the code

Eugenetanwl3881 committed:
Improve the GUI

Let's make the GUI more appealing and add personality to the GUI, Bob.
Also give credits to the code that was referred to while improving GUI.

Eugenetanwl3881 committed:
Update JavaFX FXML API version warning

Update the files to correct JavaFX version as SceneBuilder changed it
while editing the FXML files. This creates a warning if the versions are
different.

Eugenetanwl3881 committed:
Merge branch 'branch-A-BetterGUI'

Eugenetanwl3881 committed:
Add User Guide

There is no user guide for Bob the Task Manager Chatbot.
Let's set up a UserGuide for the application

Eugenetanwl3881 committed:
Add more Testing

Eugenetanwl3881 committed:
Prepare JAR file for release

Eugenetanwl3881 committed:
Add More Error Handling

Let's include handling more errors that may be caused by user input
